### ISSUE TITLE
Deepspeed Checkpoints Fail To Resume With PyTorch >=2.6

### ISSUE THREAD

**Issue by alexdremov** [axolotl-ai-cloud/axolotl#2610]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Hi!

Due to recent PyTorch's change in default behavior for loading checkpoints in weight-only format, `deepspeed` version must be bumped up. 

That's where `weights_only=False` appears in newer versions:
https://github.com/deepspeedai/DeepSpeed/blame/ee492c30a716474578d6d42115f2698ab131ab79/deepspeed/runtime/checkpoint_engine/torch_checkpoint_engine.py#L28

### Current behaviour

Fail to load checkpoint with `weights_only=True` as default

### Steps to reproduce

Try to load deepspeed checkpoint 

### Config yaml

```yaml

```

### Possible solution

Bump up deepspeed version to the most recent

### Which Operating Systems are you using?

- [x] Linux
- [x] macOS
- [x] Windows

### Python Version

3.11

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey! Thanks for the report.

Have you given any of the recent deepspeed a try? Reason we locked to the current was some gradient_checkpoint issue causing it to use more VRAM in newer releases.

Have you noticed any difference on that front?

**Comment by alexdremov** [comment]
> Hey! Thanks for the report.
> 
> Have you given any of the recent deepspeed a try? Reason we locked to the current was some gradient_checkpoint issue causing it to use more VRAM in newer releases.
> 
> Have you noticed any difference on that front?

I did not measure memory consumption, but I did notice that some additional fixes are needed (at least torch compile handling for example)

**Comment by NanoCode012** [comment]
Could you elaborate on the torch compile part?

**Comment by alexdremov** [comment]
> Hey! Thanks for the report.
> 
> Have you given any of the recent deepspeed a try? Reason we locked to the current was some gradient_checkpoint issue causing it to use more VRAM in newer releases.
> 
> Have you noticed any difference on that front?

I don't see compile configuration in the most recent deepspeed docs. Seems like it was removed, so usual torch compile can be used

**Comment by alexdremov** [comment]
Also, DeepCompile was released, so configurations should be patched

**Comment by NanoCode012** [comment]
> Also, DeepCompile was released, so configurations should be patched

Hey, if you have the bandwidth, would you like to give it a try then make a PR on this? I read on the docs on this a while back and it theoretically should only require setting the `compile` section.

---

Back to your original issue, did the newer deepspeed version fixed it for you? If yes, which version?

**Comment by alexdremov** [comment]
> > Also, DeepCompile was released, so configurations should be patched
> 
> Hey, if you have the bandwidth, would you like to give it a try then make a PR on this? I read on the docs on this a while back and it theoretically should only require setting the `compile` section.
> 
> Back to your original issue, did the newer deepspeed version fixed it for you? If yes, which version?

Sure, I'll try to draft a PR with tests for different configurations. 

I am using the most recent version. Probably, there's a lower version that fixes checkpoints issue, but there was no point for me to try them.

**Comment by winglian** [comment]
Related https://github.com/huggingface/transformers/pull/37785

**Comment by winglian** [comment]
also related  https://github.com/huggingface/transformers/pull/36991 and https://github.com/huggingface/transformers/pull/36991

**Comment by NanoCode012** [comment]
> Sure, I'll try to draft a PR with tests for different configurations.
> 
> I am using the most recent version. Probably, there's a lower version that fixes checkpoints issue, but there was no point for me to try them.

Thanks!

**Comment by ved1beta** [comment]
Trained SmolLM2-135M on 2xA100 with deepspeed_configs/zero2.json, save_steps: 2; checkpoint-2 contains global_step2/bf16_zero_pp_rank_{0,1}_mp_rank_00_optim_states.pt. Copied that checkpoint alone into a clean output dir and resumed with resume_from_checkpoint: it logged exactly 4 optimizer steps starting at epoch 0.003067
